Source/DafnyLanguageServer/Workspace/ITelemetryPublisher.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace {
+  /// <summary>
+  /// Implementations of this interface are responsible to publish telemetry events
+  /// of a <see cref="DafnyDocument"/> to the LSP client.
+  /// </summary>
+  public interface ITelemetryPublisher {
+    protected enum TelemetryEventKind {
+      UpdateComplete,
+      UnhandledException
+    }
+
+    /// <summary>
+    /// Publish a telemetry event.
+    /// </summary>
+    /// <param name="kind">The kind of telemetry event.</param>
+    /// <param name="evt">The telemetry event.</param>
+    protected void PublishTelemetry(TelemetryEventKind kind, object? evt);
```

Comment:
It feels super strange to me to have `protected` members in an interface. Although that's a feature introduced with C# 8. I've just noticed this because it's rather unusual to declare accessibility in interfaces since all members are public by default.
However, I'm not a fan of abbreviations. Why not use `payload` or something like that? The event kind is determined by the `kind` parameter, so it's not actually a full event object from my perspective.

Reply:
Good points, fixed!

Reply:
Actually, I fixed the parameter name, but not the interface part.  Do you think I should make it an abstract class instead of an interface?  I'm using this interface + implementation pattern because that's the style of the codebase, but I didn't give it that much thought.

I made the method protected because it shouldn't be exposed to clients — the only thing that they should call are the higher-level methods.